core/trino-main/src/main/java/io/trino/connector/CoordinatorDynamicCatalogManager.java

```diff
@@ -286,6 +289,100 @@ public void createCatalog(CatalogName catalogName, ConnectorName connectorName,
+                    handle -> catalogFactory.createCatalog(newCatalogProperties));
+            catalogStore.addOrReplaceCatalog(newCatalogProperties);
+            activeCatalogs.put(newCatalogName, newCatalog.getCatalog());
+            catalogStore.removeCatalog(catalogName);
+            activeCatalogs.remove(catalogName);
+        }
+        finally {
+            catalogsUpdateLock.unlock();
+        }
+    }
+
+    @Override
+    public void alterCatalog(CatalogName catalogName, Map<String, Optional<String>> properties)
+    {
+        requireNonNull(catalogName, "catalogName is null");
+
+        catalogsUpdateLock.lock();
+        try {
+            checkState(state != State.STOPPED, "ConnectorManager is stopped");
+
+            if (!activeCatalogs.containsKey(catalogName)) {
+                throw new TrinoException(NOT_FOUND, format("Catalog '%s' does not exist", catalogName));
+            }
+
+            CatalogHandle catalogHandle = activeCatalogs.get(catalogName).getCatalogHandle();
+            CatalogProperties oldCatalogProperties = allCatalogs.get(catalogHandle).getCatalogProperties().orElseThrow();
+            CatalogProperties newCatalogProperties = catalogStore.createCatalogProperties(
+                    catalogName,
+                    oldCatalogProperties.connectorName(),
+                    updateProperties(oldCatalogProperties, properties));
+
+            // get or create catalog for the handle
+            CatalogConnector newCatalog = allCatalogs.computeIfAbsent(
+                    newCatalogProperties.catalogHandle(),
+                    handle -> catalogFactory.createCatalog(newCatalogProperties));
+            catalogStore.addOrReplaceCatalog(newCatalogProperties);
+            activeCatalogs.put(catalogName, newCatalog.getCatalog());
+
+            log.debug("Updated catalog: " + catalogName);
+        }
+        finally {
+            catalogsUpdateLock.unlock();
+        }
+    }
+
+    private static Map<String, String> updateProperties(CatalogProperties existingProperties, Map<String, Optional<String>> changedProperties)
+    {
+        ImmutableSet.Builder<String> removedPropertiesBuilder = ImmutableSet.builder();
+        ImmutableMap.Builder<String, String> updatedPropertiesBuilder = ImmutableMap.builder();
+        for (Map.Entry<String, Optional<String>> property : changedProperties.entrySet()) {
+            if (property.getValue().isEmpty()) {
+                removedPropertiesBuilder.add(property.getKey());
+            }
+            else {
+                updatedPropertiesBuilder.put(property.getKey(), property.getValue().get());
+            }
+        }
+        Set<String> removedProperties = removedPropertiesBuilder.build();
+        Map<String, String> updatedProperties = updatedPropertiesBuilder.buildOrThrow();
+
+        Map<String, String> oldProperties = Maps.filterKeys(existingProperties.properties(), key -> !removedProperties.contains(key));
```

Comment:
Prefer Java streams instead of Guava's Maps.filterKeys. Guava's implementation is lazy, which can cause problems in certain situations.